widgetssdk/src/main/java/com/glia/widgets/engagement/EngagementRepositoryImpl.kt

```diff
@@ -220,13 +220,14 @@ internal class EngagementRepositoryImpl(
             unsubscribeFromEvents(it)
             resetState()
 
-            it.end { ex -> ex?.also { Logger.d(TAG, "Ending engagement failed") } }
+            //End the current engagement only when it is ended by the visitor.
+            if (endedBy == EndedBy.VISITOR) {
+                it.end { ex -> ex?.also { Logger.d(TAG, "Ending engagement failed") } }
+            }
```

Comment:
So, the endEngagement() will not end engagement if it's called programmatically (not be visitor), do I understand correctly? Could you please describe the reason for this change?

Reply:
I like the idea, but I think that's not in the scope of this fix.
I will add tests to this PR, so we can include this fix to 3.1.1 release.
Sine the suggested improvements will touch multiple layers with business logic, I'll create separate PR for that, this way we will test changes until the next release.